engines/immortal/compression.cpp

```diff
@@ -0,0 +1,288 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "immortal/immortal.h"
+
+/* Decompression:
+ * This decompression algorithm follows the source assembly very closely,
+ * which is itself a modification to LZW (a derivitive of LZ78).
+ * In: Source data as File, size of data
+ * Out: Pointer to uncompressed data as SeekableReadStream
+ */
+namespace Immortal {
+
+Common::SeekableReadStream *ImmortalEngine::unCompress(Common::File *src, int srcLen) {
+	/* Note: this function does not seek() in the file, which means
+	 * that if there is a header on the data, the expectation is that
+	 * seek() was already used to move past the header before this function.
+	 */
+
+	// If the source data has no length, we certainly do not want to decompress it
+	if (srcLen == 0) {
+		return nullptr;
+	}
+
+	/* This will be the dynamically re-allocated writeable memory stream.
+	 * We do not want it to be deleted from scope, as this location is where
+	 * the readstream being returned will point to.
+	 */
+	Common::MemoryWriteStreamDynamic dstW(DisposeAfterUse::NO);
+
+	// The 20k bytes of memory that compression gets allocated to work with for the dictionary and the stack of chars
+	uint16 start[0x4000];                   // Really needs a better name, remember to do this future me
+	uint16   ptk[0x4000];                   // Pointer To Keys? Also needs a better name
+	  byte stack[0x4000];                   // Stack of chars to be stored
+
+	// These are the main variables we'll need for this
+	uint16 findEmpty;
+	uint16 code;                            // Needs to be ASL to index with
+	uint16 inputCode;
+	uint16 finalChar;
+	uint16 myCode;                          // Silly name is silly
+	uint16 oldCode;
+	uint16 index;                           // The Y register was used to index the byte array's, this will sort of take its place
+	uint16 evenOdd  = 0;
+	uint16 topStack = 0;
+
+	byte outByte;                           // If only we could SEP #$20 like the 65816
+
+	setupDictionary(start, ptk, findEmpty); // Clear the dictionary and also set findEmpty to 8k
+	bool carry = true;                      // This will represent the carry flag so we can make this a clean loop
+
+	code = getInputCode(carry, src, srcLen, evenOdd); // Get the first code
+	if (carry == false) {
+		return nullptr;                     // This is essentially the same as the first error check, but the source returns an error code and didn't even check it here so we might as well
+	}
+
+	finalChar = code;
+	  oldCode = code;
+	   myCode = code;
+
+	outByte = code & kMaskLow;
+	dstW.writeByte(outByte);                // Take just the lower byte and write it the output
+
+	// :nextcode
+	while (carry == true) {
+
+		code = getInputCode(carry, src, srcLen, evenOdd); // Get the next code
+		if (carry == true) {
+
+				index = code << 1;
```

Comment:
Indentation is all off here. We align code by left edge, not by the assignment operators.